track/align_guidescope.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+
+# TODO: Update this docstring
+"""
+Plate solve a frame from guidscope camera and use the position of a particular named star to
+apply a guidescope alignment offset to the default mount model.
+
+It is assumed that the named star is already centered in the main OTA camera when this program is
+run. This might be accomplished by running the main `track` program in sensor fusion mode using
+the main OTA camera for guiding rather than the guidescope camera (which is used here). Both
+programs can run concurrently since they access different cameras.
+"""
+
+import atexit
+import os
+import sys
+import signal
+import subprocess
+import time
+from astropy.coordinates import SkyCoord
+import astropy.units as u
+import click
+import ephem
+import matplotlib.pyplot as plt
+import track
+from track import cameras
+from track.config import CONFIG_PATH
+from track.mounts import MeridianSide
+from track.model import (
+    save_default_param_set,
+    load_stored_param_set,
+    apply_guide_cam_alignment_error
+)
+
+
+SHUTDOWN_TIMEOUT_S = 2.0
+
+
+# FIXME: To shut down the subprocess cleanly, we need to send it SIGINT once and *only* once. Since
+# the subprocess is part of the same process group, when we do CTRL-C at the terminal *all*
+# processes in that group will receive SIGINT, and thus there is no need to send SIGINT to the
+# subprocess a second time under these conditions (and in fact problems arise if it gets SIGINT
+# more than once for reasons I don't fully understand). However, when this program ends for reasons
+# other than receiving CTRL-C, we *do* need to send SIGINT to the subprocess.
+
+
+def terminate_subprocess(process: subprocess.Popen) -> None:
+    """Terminate a running subprocess"""
+    print('Terminating subprocess...', end='', flush=True)
+    process.send_signal(signal.SIGINT)
+    time_sigint = time.perf_counter()
+    while process.poll() is None:
+        if time.perf_counter() - time_sigint > SHUTDOWN_TIMEOUT_S:
+            print('Subprocess shutdown timeout; resorting to SIGKILL...', end='', flush=True)
+            process.kill()
+            break
+        time.sleep(0.1)
+    print('done.')
+
+
+def main():
+    """See module docstring at the top of this file."""
+    parser = track.ArgParser()
+    parser.add_argument('star', help='Name of bright star')
+    parser.add_argument(
+        '--main-camera-cfg',
+        help='Configuration file with settings for the main OTA camera',
+        default=os.path.join(CONFIG_PATH, 'camera_main.cfg')
+    )
+    parser.add_argument(
+        '--meridian-side',
+        help='side of meridian for equatorial mounts to prefer',
+        required=True,
+        choices=tuple(m.name.lower() for m in MeridianSide),
+    )
+    cameras.add_program_arguments(parser, profile='align')
+    args = parser.parse_args()
+
+    # Run the `track` program as a subprocess. The intent is for this process to point the mount at
```

Comment:
- [ ] The spiral search starts immediately, but it really shouldn't start until the mount has arrived at the predicted location of the star. Figure out how to cause the spiral search to wait. Perhaps the easiest way is to just run `track` prior to this in a configuration that blindly points to the star location and then quits. Then when `track` is started here the mount should already be in the vicinity of the star and it will be appropriate for the spiral search to start immediately.